web3.js/src/connection.ts

```diff
@@ -5098,11 +5113,31 @@ export class Connection {
                 if (!isCurrentConnectionStillActive()) {
                   return;
                 }
-                // TODO: Maybe add an 'errored' state or a retry limit?
-                this._subscriptionsByHash[hash] = {
-                  ...subscription,
-                  state: 'pending',
-                };
+
+                const isErrorSubscription =
+                  isErrorStatefulSubscription(subscription);
+
+                if (
+                  isErrorSubscription &&
+                  subscription.retryCount + 1 > SUBSCRIPTION_ERROR_RETRY_LIMIT
+                ) {
+                  console.error(
+                    `${method} error, max retries reached`,
+                    args,
+                    e instanceof Error ? e.message : `${e}`,
+                  );
+
+                  delete this._subscriptionsByHash[hash];
+                } else {
+                  this._subscriptionsByHash[hash] = {
+                    ...subscription,
+                    retryCount: isErrorSubscription
+                      ? subscription.retryCount + 1
+                      : 0,
```

Comment:
Appreciate the response. I agree the websocket issue just needs to be fixed but this issue is crashing our serverless functions.

Reply:
Did you get a chance to check out #27859? That could very well turn out to be the entire problem.

Reply:
Yeah I did, I don't think `Remove HTTP keep-alive functionality completely from Web3js so it closes sockets as soon as the client gets a response.` is a possible outcome because of subscriptions right? I've seen "socket hang up" before though for sure.